### PR TITLE
Vesting and team payments contracts

### PR DESCRIPTION
**Pull request by princetonbishop**
# Description
What does this PR solve?

Fixes # (issue)

# Checklist
- [ ] Code follows the style guide
- [ ] I have performed a self-review of my own code
- [ ] New and existing tests pass locally
- [ ] This PR is targeting the correct branch

### CODE REVIEW

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        return _vestingIds.contains(_id);
+    }
+
+    /**
+     * @notice Check if vesting is revoked
+     * @param _vestingId Vesting Id
+     * @return Bool
+     */
+    function isVestingRevoked(bytes32 _vestingId) external view override returns (bool) {
+        if(!vestingIdExists(_vestingId)) { return true; }
```
Comment (frontier159) [inline protocol/contracts/templegold/TempleGoldVesting.sol:192]: Is this check semantically incorrect? If an id hasn't been created, that doesn't mean its been revoked, it just mean it doesn't exist yet.

I think this line can just be removed?

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+
+    /// @notice Total TGLD vested and unclaimed
+    uint256 public override totalVestedAndUnclaimed;
+    /// @notice Schdules for recipients
+    mapping(bytes32 id => VestingSchedule schedule) public schedules;
+      /// @notice Recipient vesting counts for generating IDs. An account can have multiple vesting schedules
+    mapping(address => uint256) public holdersVestingCount;
+    /// @notice Vesting Ids
+    EnumerableSet.Bytes32Set private _vestingIds;
```
Comment (frontier159) [inline protocol/contracts/templegold/TempleGoldVesting.sol:32]: I wonder if this should be renamed to `_activeVestingIds`

Since when they get revoked they get removed from here (and he order of the items in the enumerable set will change from that)

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    ) public pure override returns (bytes32) {
+        return keccak256(abi.encodePacked(holder, index));
+    }
+
+    /**
+     * @notice Check if vesting Id exists
+     * @param _id Vesting Id
+     * @return Bool
+     */
+     function vestingIdExists(bytes32 _id) public view returns (bool) {
```
Comment (frontier159) [inline protocol/contracts/templegold/TempleGoldVesting.sol:182]: A better name might be `isActiveVestingId(_id)`, since revoked id's still 'exist' in the schedules map but it's just no longer active because it got revoked

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+     * @return VestingSchedule
+     */
+    function getLastVestingScheduleForHolder(
+        address _recipient
+    ) external view override returns (VestingSchedule memory) {
+        return
+            schedules[
+                computeVestingScheduleIdForAddressAndIndex(
+                    _recipient,
+                    holdersVestingCount[_recipient] - 1
```
Comment (frontier159) [inline protocol/contracts/templegold/TempleGoldVesting.sol:217]: This will revert with an underflow panic if there are no schedules added for this account. Is that the intention?

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+    function getSchedule(bytes32 _vestingId) external view override returns (VestingSchedule memory schedule) {
+        schedule = schedules[_vestingId];
+    }
+
+    /**
+     * @notice Returns the last vesting schedule for a given holder address.
+     * @param _recipient Recipient address
+     * @return VestingSchedule
+     */
+    function getLastVestingScheduleForHolder(
```
Comment (frontier159) [inline protocol/contracts/templegold/TempleGoldVesting.sol:210]: I don't see tests for these...need to make sure you have tests for **everything**, and that you cover all edge cases please.

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+     * @notice Get total vested amount at current block timestamp for vesting Id
+     * @param _vestingId Vesting Id
+     * @return Vested amount 
+     */
+    function getTotalVestingAtCurrentTime(bytes32 _vestingId) external view override returns (uint256) {
+        VestingSchedule memory _schedule = schedules[_vestingId];
+        if (_schedule.amount == 0) { return 0; }
+        if (block.timestamp <= _schedule.start) { return 0; }
+        if (_schedule.revoked) { return 0; }
+        uint32 _elapsed = _getElapsedTime(_schedule.start, uint32(block.timestamp), _schedule.duration);
```
Comment (frontier159) [inline protocol/contracts/templegold/TempleGoldVesting.sol:245]: Can use `_calculateTotalVestedAt()`?

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+    /**
+     * @notice Get vesting summary
+     * @param _ids Vesting Ids
+     * @param _from From timestamp
+     * @param _to To timestamp
+     * @return summary Vesting summary 
+     */
+    function getVestingSummary(
+        bytes32[] memory _ids,
+        uint32 _from,
```
Comment (frontier159) [inline protocol/contracts/templegold/TempleGoldVesting.sol:258]: Are the `_from` and `_to` very useful, wouldn't the client just filter out those after?

Also `_to` is a little weird because there could be a bunch of vested but not yet claimed from past schedules that you wouldn't want to drop out of the summary?

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+        return _calculateTotalVestedAt(_schedule, _at);
+    }
+
+    function _calculateReleasableAmount(
+          VestingSchedule storage _schedule
+    ) private view returns (uint256) {
+        if ((block.timestamp <= _schedule.cliff) || _schedule.revoked) {
+            return 0;
+        } else {
+            uint256 _vested = TempleMath.mulDivRound(_schedule.amount, block.timestamp - _schedule.start, _schedule.duration, false);
```
[the remark this answers is not in the document]
Comment (frontier159) [reply protocol/contracts/templegold/TempleGoldVesting.sol:315]: Should use `_calculateTotalVestedAt()`?

**Review by frontier159 — CHANGES_REQUESTED** [review]

File: protocol/contracts/admin/TeamPayments.sol [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    /**
+     * @notice Set epoch payments
+     * @param epoch Epoch
+     * @param recipients Recipients array
+     * @param amounts Amounts array to set
+     */
+    function setEpochPayments(
+        uint256 epoch,
```
[the remark this answers is not in the document]
Comment (frontier159) [reply protocol/contracts/admin/TeamPayments.sol:83]: I wonder if there should be a function create a new epoch (increment) first which is then validated within this, rather than just allow any free form uint256?

File: protocol/contracts/templegold/TempleGoldVesting.sol [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+        uint256 _length = _schedules.length;
+        VestingSchedule memory _schedule;
+        for (uint256 i; i < _length; ++i) {
+            _schedule = _schedules[i];
+            // distributed and revoked should be default values
+            if (_schedule.distributed != 0) { revert CommonEventsAndErrors.InvalidParam(); }
+            if (_schedule.revoked) { revert CommonEventsAndErrors.InvalidParam(); }
+            if (_schedule.recipient == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+            if (_schedule.start < block.timestamp) { revert CommonEventsAndErrors.InvalidParam(); }
+            if (_schedule.cliff <= _schedule.start) { revert CommonEventsAndErrors.InvalidParam(); }
```
[the remark this answers is not in the document]
Comment (princetonbishop) [reply protocol/contracts/templegold/TempleGoldVesting.sol:69]: for vesting, I don't think there will be a vesting schedule without cliff.
A 1 second or 0 second cliff (to claim immediately)  is an unlikely scenario. But if you insist, I can revert.

File: protocol/test/forge/unit/templegold/TempleGoldVesting.t.sol [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+// (tests/forge/unit/templegold/TempleGoldVesting.t.sol)
+
+import { TempleGoldCommon } from "./TempleGoldCommon.t.sol";
+import { TempleGoldVesting } from "contracts/templegold/TempleGoldVesting.sol";
+import { TempleGold } from "contracts/templegold/TempleGold.sol";
+import { FakeERC20 } from "contracts/fakes/FakeERC20.sol";
+import { ITempleGold } from "contracts/interfaces/templegold/ITempleGold.sol";
+import { ITempleGoldVesting } from "contracts/interfaces/templegold/ITempleGoldVesting.sol";
+import { CommonEventsAndErrors } from "contracts/common/CommonEventsAndErrors.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
```
[the remark this answers is not in the document]
Comment (princetonbishop) [reply protocol/test/forge/unit/templegold/TempleGoldVesting.t.sol:12]: good catch

File: protocol/contracts/admin/PaymentBase.sol [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+     * @param _amount Amount to recover
+     */
+    function _recoverToken(address _token, address _to, uint256 _amount) internal {
+        emit CommonEventsAndErrors.TokenRecovered(_to, _token, _amount);
+        IERC20(_token).safeTransfer(_to, _amount);
+    }
+
+    function _getElapsedTime(uint40 _start, uint40 _end, uint40 _duration) internal pure returns (uint40) {
+        return _end - _start > _duration ? _duration : _end - _start;
+    }
```
[the remark this answers is not in the document]
Comment (princetonbishop) [reply protocol/contracts/admin/PaymentBase.sol:47]: good one

File: protocol/test/forge/unit/admin/EpochPayments.t.sol [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+
+        skip(payment.minEpochDuration()+1);
+        payment.setEpochPayments(recipients, amounts);
+        assertEq(payment.currentEpoch(), 3);
+        assertEq(payment.epochStartTimes(3), block.timestamp);
+    }
+
+    function test_minEpochDuration() public {
+
+    }
```
[the remark this answers is not in the document]
Comment (princetonbishop) [reply protocol/test/forge/unit/admin/EpochPayments.t.sol:187]: thanks

File: protocol/contracts/interfaces/admin/IVestingPayments.sol [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+   /**
+     * @notice Get vesting summary
+     * @param _ids Vesting Ids
+     * @return summary Vesting summary 
+     */
+    function getVestingSummary(
+        bytes32[] memory _ids
+    ) external view returns (VestingSummary[] memory summary);
+
```
[the remark this answers is not in the document]
Comment (princetonbishop) [reply protocol/contracts/interfaces/admin/IVestingPayments.sol:146]: nice

File: protocol/test/forge/unit/admin/VestingPayments.t.sol [hunk 15]
```diff
@@ -0,0 +1,927 @@
...
+        totalVestedAndUnclaimed = vesting.totalVestedAndUnclaimed();
+        releasable = vesting.getReleasableAmount(_id);
+        vm.startPrank(alice);
+        vm.expectEmit(address(vesting));
+        emit Released(_id, alice, releasable);
+        vesting.release(_id);
+        uint256 distributed = _schedule.distributed + releasable;
+        _schedule = vesting.getSchedule(_id);
+        assertEq(_schedule.distributed, distributed);
+        assertEq(vesting.totalVestedAndUnclaimed(), totalVestedAndUnclaimed-releasable);
```
[the remark this answers is not in the document]
Comment (princetonbishop) [reply protocol/test/forge/unit/admin/VestingPayments.t.sol:811]: good

File: protocol/contracts/admin/PaymentBase.sol [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        /// @dev Elevated access should revoke approval from old `fundsOwner` for this contract
+        fundsOwner = _fundsOwner;
+        emit FundsOwnerSet(_fundsOwner);
+    }
+
+    /// @inheritdoc IPaymentBase
+    function recoverToken(address _token, address _to, uint256 _amount) external override onlyElevatedAccess {
+        emit CommonEventsAndErrors.TokenRecovered(_to, _token, _amount);
+        IERC20(_token).safeTransfer(_to, _amount);
+    }
```
[the remark this answers is not in the document]
Comment (princetonbishop) [reply protocol/contracts/admin/PaymentBase.sol:40]: in `safeTransfer`, there's a revert if token is address(0)

File: protocol/package.json [hunk 17]
```diff
@@ -77,7 +77,9 @@
     "local-fork:test:templev2": "hardhat run --network localhost scripts/deploys/localhost/v2/999-localhost-fork-test.ts",
     "forge-test": "forge test",
     "forge-coverage": "forge coverage",
-    "forge-coverage-html": "forge coverage --report summary --report lcov && lcov --remove lcov.info 'contracts/fakes/*' 'contracts/deprecated/*' 'test/*' 'script/*' 'node_modules/*' --output-file lcov.info --rc lcov_branch_coverage=1 && genhtml lcov.info -o report --branch-coverage --legend"
+    "forge-coverage-html-1": "forge coverage --lcov-version 2 --offline --no-match-test \"fuzz\" --no-match-path \"contracts/deprecated/**/*\" --report summary --report lcov",
+    "forge-coverage-html-2": "genhtml lcov.info -o report --exclude 'contracts/fakes/*' --exclude 'test/*' --branch-coverage --rc branch_coverage=1 --rc derive_function_end_line=0 --legend --ignore-errors inconsistent --ignore-errors category --ignore-errors unused",
+    "forge-coverage-html": "yarn forge-coverage-html-1 && yarn forge-coverage-html-2"
```
[the remark this answers is not in the document]
Comment (frontier159) [reply protocol/package.json:82]: not required, CI doesn't run this command.

File: protocol/contracts/admin/VestingPayments.sol [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+
+    function _calculateTotalVestedAt(
+        VestingSchedule memory _schedule,
+        uint40 _releaseTime
+    ) private pure returns (uint256) {
+        if (_schedule.amount == 0) { return 0; }
+        // below cliff
+        if (_releaseTime <= _schedule.cliff) { return 0; }
+        // if revoked, return total vested at time of revoke (distributed + checkpointed delta)
+        if (_schedule.revoked) { return _schedule.revokedReleasable + _schedule.distributed; }
```
Comment (frontier159) [inline protocol/contracts/admin/VestingPayments.sol:244]: Given this tweak to be `revokedReleasable + distributed`,  I think that means the line 233 doesn't need the ternary check now? It can just simplify to:

```
return _calculateTotalVestedAt(_schedule, uint40(block.timestamp)) - _schedule.distributed;
```
?

File: protocol/contracts/admin/VestingPayments.sol [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+
+        uint256 vestedNow = _getTotalVestedAtCurrentTime(_schedule);
+        uint256 unvested = _schedule.amount - vestedNow;
+        if (unvested > 0) {
+            totalVestedAndUnclaimed -= unvested;
+        }
+        // Claimable delta at revoke time
+        uint256 releasableNow = vestedNow - _schedule.distributed;
+        // Persist per-schedule claimable checkpoint
+        _schedule.revokedReleasable = releasableNow.encodeUInt128();
```
Comment (frontier159) [inline protocol/contracts/admin/VestingPayments.sol:94]: Was a specific test case added to check multiple revoke's for the same user? It doesn't look like it, but that's the exact case this was intended to fix. So that should 💯 be added in (unless i just missed it in the existing tests)

Comment (princetonbishop) [reply protocol/contracts/admin/VestingPayments.sol:94]: This was already in the existing test [previous commit](https://github.com/TempleDAO/temple/pull/1169/files#diff-5bd7314dad1cc21bbc5f3aa22b8b8b5d1a05d6175e4c4bd247b0c8b5a027a5e0R733-R739) as
```
function test_revokeVesting_invalid_id_after_revoke() public {
        test_revokeVesting_immediate_revoke_after_creation();
        // use previous index to match id
        bytes32 _id = vesting.computeVestingScheduleIdForAddressAndIndex(alice, 0);
        vm.expectRevert(abi.encodeWithSelector(CommonEventsAndErrors.InvalidParam.selector));
        vesting.revokeVesting(_id);
    }
```
Unless I misunderstood your point?
